### DISCUSSION TITLE
Make elapsed time available in the `response` event hook

### DISCUSSION THREAD

**Discussion by joaonc** [encode/httpx#2722] [General] [answered]
There are two types of elapsed time on a response:

* The current `.elapsed` that accounts for total time of request, including event hooks. This is what we have now (Related to https://github.com/encode/httpx/issues/655).
* The elapsed time of the actual request, not taking into account the event hooks, which can take a significant amount of time, depending on what you're doing. This number is currently not available.

In many scenarios, we have `request` and/or `response` event hooks that do a time consuming process, but need the time of the actual request only, w/o event hooks.  Prime example is logging the response in a `response` event hook and we want to include the elapsed time in the response.

Is it possible to add a mechanism in which:
* We can access the elapsed time w/o the time to process event hooks.
* Have that time available in the `response` hook.

**Comment by lovelydinosaur** [comment, marked as answer]
Options here...

* track your time usage inside the event hooks.
* consider the `"trace"` extension for more granualar time tracking.

**Comment by arseniiarsenii** [comment]
Here's a workaround I found:
```python3
elapsed_time_seconds: float = response.stream._timer.sync_elapsed()  # If you're using a sync client
elapsed_time_seconds: float = await response.stream._timer.async_elapsed()  # If you're using an async client
```
This is basically what `Response.elapsed` property uses under the hood. 

Using `trace` extension for this application is definitely an overkill + it's not really documented. Wasted hours going this route, sharing this solution so other people can save time.

**Reply by lovelydinosaur** [reply]
> Using trace extension for this application is definitely an overkill + it's not really documented. Wasted hours going this route, sharing this solution so other people can save time.

See also https://github.com/encode/httpx/discussions/3012
